File: dotNet/tests/VeruSuite.Api.Tests/Support.cs
using System.Net;
using System.Text;
using System.Text.Json;
using VeruSuite.Api;

namespace VeruSuite.Api.Tests;

/// <summary>One request the client sent.</summary>
public sealed record Recorded(
    string Method,
    string Path,
    ILookup<string, string> Query,
    IReadOnlyDictionary<string, string> Headers,
    string Body)
{
    public string Route => $"{Method} {Path}";

    public JsonElement Json => JsonDocument.Parse(Body).RootElement;
}

/// <summary>One response the API should give.</summary>
public sealed record Reply(int Status = 200, string? Body = null, IDictionary<string, string>? Headers = null);

/// <summary>
/// Records what the client sent and answers what it was told to.
/// </summary>
/// <remarks>
/// A handler rather than a live server: an HttpRequestMessage carries the
/// method, the URI, every header and the body, which is the whole of what the
/// API would receive. Nothing is asserted about the code calling a mock, only
/// about the request that would go out.
/// </remarks>
public sealed class FakeApi : HttpMessageHandler
{
    private readonly Queue<Reply> _replies = new();

    public FakeApi(params Reply[] replies)
    {
        foreach (var reply in replies)
        {
            _replies.Enqueue(reply);
        }
    }

    public List<Recorded> Requests { get; } = new();

    public List<TimeSpan> Slept { get; } = new();

    /// <summary>
    /// Used once the script runs out.
    /// </summary>
    /// <remarks>
    /// A null data rather than an empty array, because System.Text.Json is
    /// strict about shape: [] deserializes into a list but not into a Message,
    /// and a test that only cares about the request it sent should not have to
    /// describe the response it ignores. Null satisfies every shape, and the
    /// list-returning methods turn it back into an empty list.
    /// </remarks>
    public Reply Default { get; set; } = new(200, """{"data":null,"meta":{"request_id":"req_test"}}""");

    public Recorded Last => Requests[^1];

    /// <summary>A client pointed at this handler, with the waiting taken out.</summary>
    public VeruApiClient Client(int maxRetries = 2)
    {
        var api = new VeruApiClient("vak_live_test_secret", new VeruApiClientOptions
        {
            BaseUrl = "https://api.test.example",
            MaxRetries = maxRetries,
            HttpClient = new HttpClient(this, disposeHandler: false),
        });

        // Replaced rather than shortened: a test asserting that three attempts
        // happen should not also assert how long a coffee takes.
        api.Delay = (wait, _) =>
        {
            Slept.Add(wait);
            return Task.CompletedTask;
        };

        return api;
    }

    protected override async Task<HttpResponseMessage> SendAsync(
        HttpRequestMessage request,
        CancellationToken cancellationToken)
    {
        var uri = request.RequestUri!;
        var body = request.Content is null
            ? ""
            : await request.Content.ReadAsStringAsync(cancellationToken);

        var query = System.Web.HttpUtility.ParseQueryString(uri.Query);
        var pairs = query.AllKeys
            .Where(k => k is not null)
            .SelectMany(k => (query.GetValues(k) ?? Array.Empty<string>()).Select(v => (Key: k!, Value: v)))
            .ToLookup(p => p.Key, p => p.Value);

        var headers = request.Headers
            .Concat(request.Content?.Headers ?? Enumerable.Empty<KeyValuePair<string, IEnumerable<string>>>())
            .ToDictionary(h => h.Key.ToLowerInvariant(), h => string.Join(",", h.Value));

        Requests.Add(new Recorded(request.Method.Method, uri.AbsolutePath, pairs, headers, body));

        var reply = _replies.Count > 0 ? _replies.Dequeue() : Default;

        var response = new HttpResponseMessage((HttpStatusCode)reply.Status)
        {
            Content = new StringContent(reply.Body ?? "", Encoding.UTF8, "application/json"),
        };

        foreach (var (name, value) in reply.Headers ?? new Dictionary<string, string>())
        {
            response.Headers.TryAddWithoutValidation(name, value);
        }

        return response;
    }

    public void Enqueue(Reply reply) => _replies.Enqueue(reply);
}

public static class Bodies
{
    /// <summary>The shape every successful response arrives in.</summary>
    public static string Envelope(string data, string? nextCursor = null)
    {
        var cursor = nextCursor is null ? "" : $@",""next_cursor"":""{nextCursor}""";
        return $@"{{""data"":{data},""meta"":{{""request_id"":""req_test"",""timestamp"":""t""{cursor}}}}}";
    }

    /// <summary>The shape every refusal arrives in.</summary>
    public static string Refusal(string code, string message = "no", string? details = null)
    {
        var extra = details is null ? "" : $@",""details"":{details}";
        return $@"{{""error"":{{""code"":""{code}"",""message"":""{message}"",""request_id"":""req_test""{extra}}}}}";
    }
}

File: python/tests/support.py
"""A real HTTP server for the tests to talk to.

A recording server rather than a patched transport. The point of a client
library is what goes on the wire, and a mock that stands in for urllib asserts
that the code calls the mock. This asserts the request the API would actually
receive, which is the only thing that can be wrong.
"""

from __future__ import annotations

import json
import threading
from dataclasses import dataclass, field
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from typing import Any, Dict, List, Optional, Tuple
from urllib.parse import parse_qs, urlparse

from veruapis import VeruApi


@dataclass
class Recorded:
    """One request the server received."""

    method: str
    path: str
    query: Dict[str, List[str]]
    headers: Dict[str, str]
    body: Optional[Any]

    @property
    def route(self) -> str:
        return f"{self.method} {self.path}"


@dataclass
class Reply:
    """One response the server should give."""

    status: int = 200
    body: Any = None
    headers: Dict[str, str] = field(default_factory=dict)

    # When set, the body is sent verbatim rather than as JSON, so a test can
    # send something that is not an envelope at all.
    raw: Optional[bytes] = None


class FakeAPI:
    """A server that records what it was asked and answers what it was told."""

    def __init__(self, *replies: Reply) -> None:
        self.replies: List[Reply] = list(replies)
        self.requests: List[Recorded] = []

        # The reply used once the script runs out. An empty envelope keeps a
        # test that only cares about the request from having to describe one.
        self.default = Reply(200, {"data": [], "meta": {"request_id": "req_test"}})

        recorder = self

        class Handler(BaseHTTPRequestHandler):
            protocol_version = "HTTP/1.1"

            def log_message(self, *args: Any) -> None:
                pass  # pragma: no cover - silence the test output

            def _handle(self) -> None:
                parsed = urlparse(self.path)

                length = int(self.headers.get("Content-Length") or 0)
                raw = self.rfile.read(length) if length else b""

                # Almost every request is JSON, and one is not: a part of a
                # resumable upload carries the file's bytes. Recording those as
                # bytes rather than failing to parse them is what lets a test
                # assert that the client did not JSON-encode a file.
                if not raw:
                    body = None
                else:
                    try:
                        body = json.loads(raw)
                    except (json.JSONDecodeError, UnicodeDecodeError):
                        # UnicodeDecodeError as well as the JSON one: a part of
                        # an upload is arbitrary bytes, and most files are not
                        # valid UTF-8 at all.
                        body = raw

                recorder.requests.append(
                    Recorded(
                        method=self.command,
                        path=parsed.path,
                        query=parse_qs(parsed.query, keep_blank_values=True),
                        headers={k.lower(): v for k, v in self.headers.items()},
                        body=body,
                    )
                )

                reply = recorder.replies.pop(0) if recorder.replies else recorder.default

                payload = reply.raw
                if payload is None:
                    payload = b"" if reply.body is None else json.dumps(reply.body).encode()

                self.send_response(reply.status)
                for name, value in reply.headers.items():
                    self.send_header(name, value)
                self.send_header("Content-Type", "application/json")
                self.send_header("Content-Length", str(len(payload)))
                self.end_headers()
                if payload:
                    self.wfile.write(payload)

            do_GET = do_POST = do_PUT = do_PATCH = do_DELETE = _handle

        self._server = ThreadingHTTPServer(("127.0.0.1", 0), Handler)
        self._thread = threading.Thread(
            # serve_forever polls, and the default 500ms is paid on every
            # shutdown. With a server per test that is most of the suite.
            target=lambda: self._server.serve_forever(poll_interval=0.01),
            daemon=True,
        )
        self._thread.start()

    @property
    def base_url(self) -> str:
        host, port = self._server.server_address[:2]
        return f"http://{host}:{port}"

    def client(self, **options: Any) -> VeruApi:
        """A client pointed at this server, with the waiting taken out.

        Retry delays are replaced rather than shortened: a test asserting that
        three attempts happen should not also assert how long a coffee takes.
        """
        api = VeruApi("vak_live_test_secret", base_url=self.base_url, **options)

        self.slept: List[float] = getattr(self, "slept", [])
        api._sleep = self.slept.append
        return api

    @property
    def last(self) -> Recorded:
        return self.requests[-1]

    def close(self) -> None:
        self._server.shutdown()
        self._server.server_close()
        self._thread.join(timeout=5)

    def __enter__(self) -> "FakeAPI":
        return self

    def __exit__(self, *exc: Any) -> None:
        self.close()


def envelope(data: Any, **meta: Any) -> Dict[str, Any]:
    """The shape every successful response arrives in."""
    return {"data": data, "meta": {"request_id": "req_test", "timestamp": "t", **meta}}


def refusal(code: str, message: str = "no", **extra: Any) -> Dict[str, Any]:
    """The shape every refusal arrives in."""
    return {"error": {"code": code, "message": message, "request_id": "req_test", **extra}}
